File: bench/dijkstra/dijkstra.c
/* Dijkstra head-to-head, C port. Same LCG graph + checksum as the tycho/Go ports.
 * Adjacency list as a per-node grown Edge array (Edge** + lengths) -- the C analogue
 * of tycho's `[[Edge]]`. A binary min-heap with lazy deletion. Distances are
 * tie-break independent, so any correct shortest-path reproduces the checksum. */
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

typedef struct { int to; long w; } Edge;

typedef struct { long d; int node; } HItem;
typedef struct { HItem *a; long len, cap; } Heap;

static void hpush(Heap *h, long d, int node) {
    if (h->len == h->cap) { h->cap = h->cap ? h->cap * 2 : 1024; h->a = realloc(h->a, (size_t)h->cap * sizeof(HItem)); }
    long i = h->len++;
    h->a[i].d = d; h->a[i].node = node;
    while (i > 0) { long p = (i - 1) / 2; if (h->a[p].d <= h->a[i].d) break; HItem t = h->a[p]; h->a[p] = h->a[i]; h->a[i] = t; i = p; }
}
static HItem hpop(Heap *h) {
    HItem top = h->a[0];
    h->a[0] = h->a[--h->len];
    long i = 0;
    for (;;) {
        long l = 2 * i + 1, r = 2 * i + 2, s = i;
        if (l < h->len && h->a[l].d < h->a[s].d) s = l;
        if (r < h->len && h->a[r].d < h->a[s].d) s = r;
        if (s == i) break;
        HItem t = h->a[i]; h->a[i] = h->a[s]; h->a[s] = t; i = s;
    }
    return top;
}

int main(void) {
    int n = 300000, deg = 4;
    uint64_t state = 88172645463325252ULL;
    Edge **adj = calloc((size_t)n, sizeof(Edge *));
    int *cnt = calloc((size_t)n, sizeof(int));
    for (int u = 0; u < n; u++) {
        adj[u] = malloc((size_t)deg * sizeof(Edge));
        for (int k = 0; k < deg; k++) {
            state = state * 6364136223846793005ULL + 1442695040888963407ULL;
            int to = (int)((state & 1073741823ULL) % (uint64_t)n);
            state = state * 6364136223846793005ULL + 1442695040888963407ULL;
            long w = (long)((state & 1073741823ULL) % 100ULL) + 1;
            adj[u][k].to = to; adj[u][k].w = w;
        }
        cnt[u] = deg;
    }
    const long INF = 4000000000000000000L;
    long *dist = malloc((size_t)n * sizeof(long));
    for (int i = 0; i < n; i++) dist[i] = INF;
    dist[0] = 0;
    Heap h = {0};
    hpush(&h, 0, 0);
    while (h.len > 0) {
        HItem c = hpop(&h);
        long d = c.d; int u = c.node;
        if (d > dist[u]) continue;
        for (int j = 0; j < cnt[u]; j++) {
            Edge e = adj[u][j];
            long nd = d + e.w;
            if (nd < dist[e.to]) { dist[e.to] = nd; hpush(&h, nd, e.to); }
        }
    }
    long sum = 0, reach = 0;
    for (int i = 0; i < n; i++) if (dist[i] < INF) { sum += dist[i]; reach++; }
    printf("%ld %ld\n", reach, sum);
    return 0;
}

File: bench/dijkstra/dijkstra.go
// Dijkstra head-to-head, Go port. Same LCG graph + checksum as the tycho/C ports.
// Adjacency as [][]Edge; a hand-built binary min-heap with lazy deletion (the same
// shape as the other ports, not container/heap, to keep the comparison about the data).
package main

import "fmt"

type Edge struct {
	to int
	w  int64
}
type HItem struct {
	d    int64
	node int
}

func main() {
	n, deg := 300000, 4
	var state uint64 = 88172645463325252
	adj := make([][]Edge, n)
	for u := 0; u < n; u++ {
		adj[u] = make([]Edge, deg)
		for k := 0; k < deg; k++ {
			state = state*6364136223846793005 + 1442695040888963407
			to := int((state & 1073741823) % uint64(n))
			state = state*6364136223846793005 + 1442695040888963407
			w := int64((state&1073741823)%100) + 1
			adj[u][k] = Edge{to, w}
		}
	}
	const INF int64 = 4000000000000000000
	dist := make([]int64, n)
	for i := range dist {
		dist[i] = INF
	}
	dist[0] = 0
	heap := make([]HItem, 0, 1024)
	heap = append(heap, HItem{0, 0})
	hpush := func(d int64, node int) {
		heap = append(heap, HItem{d, node})
		i := len(heap) - 1
		for i > 0 {
			p := (i - 1) / 2
			if heap[p].d <= heap[i].d {
				break
			}
			heap[p], heap[i] = heap[i], heap[p]
			i = p
		}
	}
	hpop := func() HItem {
		top := heap[0]
		heap[0] = heap[len(heap)-1]
		heap = heap[:len(heap)-1]
		i := 0
		for {
			l, r, s := 2*i+1, 2*i+2, i
			if l < len(heap) && heap[l].d < heap[s].d {
				s = l
			}
			if r < len(heap) && heap[r].d < heap[s].d {
				s = r
			}
			if s == i {
				break
			}
			heap[i], heap[s] = heap[s], heap[i]
			i = s
		}
		return top
	}
	for len(heap) > 0 {
		c := hpop()
		d, u := c.d, c.node
		if d > dist[u] {
			continue
		}
		for _, e := range adj[u] {
			nd := d + e.w
			if nd < dist[e.to] {
				dist[e.to] = nd
				hpush(nd, e.to)
			}
		}
	}
	var sum, reach int64
	for i := 0; i < n; i++ {
		if dist[i] < INF {
			sum += dist[i]
			reach++
		}
	}
	fmt.Println(reach, sum)
}
